src/storage/txn/actions/commit.rs

```diff
@@ -403,4 +441,83 @@ pub mod tests {
         must_not_have_write(&mut engine, k2, 20);
         must_not_have_write(&mut engine, k2, 10);
     }
+
+    #[test]
+    fn test_commit_txn_lock_not_found() {
+        let mut engine = TestEngineBuilder::new().build().unwrap();
+        let (start_ts, commit_ts) = (TimeStamp::new(10), TimeStamp::new(20));
+        let (k1, k2, k3) = (b"k1", b"k2", b"k3");
+        must_prewrite_put(&mut engine, k1, b"v1", k1, start_ts);
+        must_rollback(&mut engine, k1, start_ts, false);
+        must_prewrite_put(&mut engine, k2, b"v2", k1, start_ts);
+        must_rollback(&mut engine, k2, start_ts, false);
+
+        // Do not collect mvcc case
+        for k in [k1, k2, k3] {
+            let key_str = String::from_utf8_lossy(k);
+            let role = if k == k1 {
+                // CommitRole is `Primary` and lock rollback.
+                // It is an expected case (rollback by other txn) and should not collect mvcc.
+                Some(CommitRole::Primary)
+            } else {
+                // When cannot determine the commit role, do not collect mvcc because it may be
+                // the primary commit and it is an expected case.
+                None
+            };
+            let err = must_err(&mut engine, k, start_ts, commit_ts, role);
+            assert_matches!(
+                err,
+                Error(box ErrorInner::TxnLockNotFound {
+                    start_ts,
+                    commit_ts,
+                    key,
+                    mvcc_info,
+                }) if {
+                    assert_eq!(key, k.to_vec());
+                    assert_eq!(start_ts,  10.into(), "key: {}", key_str);
+                    assert_eq!( commit_ts,  20.into(), "key: {}", key_str);
+                    assert_matches!(mvcc_info.clone(), None, "key: {}", key_str);
+                    true
+                }
+            );
+        }
+
+        // Should collect mvcc for debugging when secondary commit returns an error
+        // TxnLockNotFound.
+        // - k2, lock rollback.
+        // - k3, no lock.
```

Comment:
// - k4, lock exists but it is pessimistic lock type, which is also an unexpected situation.

Reply:
OK, I'll add this case